evm/src/trace/identifier/etherscan.rs

```diff
@@ -0,0 +1,97 @@
+use super::TraceIdentifier;
+use ethers::{
+    abi::{Abi, Address},
+    etherscan,
+    types::Chain,
+};
+use futures::stream::{self, StreamExt};
+use std::{borrow::Cow, path::PathBuf};
+use tokio::time::{sleep, Duration};
+use tracing::{trace, warn};
+
+/// A trace identifier that tries to identify addresses using Etherscan.
+pub struct EtherscanIdentifier {
+    /// The Etherscan client
+    client: Option<etherscan::Client>,
+}
+
+impl EtherscanIdentifier {
+    pub fn new(
+        chain: Option<impl Into<Chain>>,
+        etherscan_api_key: String,
+        cache_path: Option<PathBuf>,
+    ) -> Self {
+        if let Some(cache_path) = &cache_path {
+            if let Err(err) = std::fs::create_dir_all(cache_path.join("sources")) {
+                warn!("could not create etherscan cache dir: {:?}", err);
+            }
+        }
+
+        Self {
+            client: chain
+                .map(|chain| {
+                    etherscan::Client::new_cached(chain.into(), etherscan_api_key, cache_path).ok()
+                })
+                .flatten(),
+        }
+    }
+}
+
+impl TraceIdentifier for EtherscanIdentifier {
+    fn identify_addresses(
+        &self,
+        addresses: Vec<(&Address, Option<&Vec<u8>>)>,
+    ) -> Vec<(Address, Option<String>, Option<String>, Option<Cow<Abi>>)> {
+        if let Some(client) = &self.client {
+            let stream = stream::iter(addresses.into_iter().map(futures::future::ready))
+                .buffered(10)
+                .filter_map(|(addr, _)| {
+                    let client = client.clone();
+                    async move {
+                        let mut i = 0;
+                        trace!("requesting etherscan info for contract {addr}");
+                        loop {
```

Comment:
This is a pretty naive rate limit handling algo but it should work under the assumption that we don't request too many addresses, which I think is a fair assumption to make

Reply:
Also another assumption: We assume that the rate limit is 5 req/s